File: DisasterServer/Entities/Ring.cs
using DisasterServer.Maps;
using DisasterServer.Session;
using DisasterServer.State;

namespace DisasterServer.Entities;

public class Ring : Entity
{
	public ushort ID = 1;

	public byte IID;

	public bool IsRedRing;

	private static readonly Random _rand = new Random();

	public override TcpPacket? Spawn(Server server, Game game, Map map)
	{
		ID = map.RingIDs++;
		IsRedRing = map.CanSpawnRedRings() && _rand.Next(100) <= 10;
		return new TcpPacket(PacketType.SERVER_RING_STATE, (byte)0, IID, ID, IsRedRing);
	}

	public override TcpPacket? Destroy(Server server, Game game, Map map)
	{
		map.FreeRingID(IID);
		return new TcpPacket(PacketType.SERVER_RING_STATE, (byte)1, IID, ID);
	}

	public override UdpPacket? Tick(Server server, Game game, Map map)
	{
		return null;
	}
}

File: disasterserver/Entities/Ring.cpp
#include "Ring.hpp"

#include "Util/Packet.hpp"

using namespace DisasterServer;
using namespace DisasterServer::Entities;

Ring::Ring(entityId id, Server &server, GameState &state, const Vector2 &position, const bool red) :
    Entity(id, server, state, "cring", position), red(red) {}

Ring::~Ring() = default;

bool Ring::init() {

    Packet pack(PacketType::SERVER_RING_STATE);
    pack.write<uint8_t>(2);
    pack.writeVector2(position);
    pack.write<uint8_t>(rid);
    pack.write<entityId>(id);
    pack.write<uint8_t>(red);
    pack.sendBroadcast(server);

    return true;
}

bool Ring::uninit() {

    Packet pack(PacketType::SERVER_RING_STATE);
    pack.write<uint8_t>(1);
    pack.write<uint8_t>(rid);
    pack.write<entityId>(id);
    pack.sendBroadcast(server);

    return true;
}
